playlist.c

```diff
@@ -491,66 +642,336 @@ size_t playlist_size(playlist_t *playlist)
+            /* handle any top-level playlist metadata here */
+            /*RARCH_LOG("found meta: %s = %s\n", pCtx->current_meta_string, pValue);*/
+
+            free(pCtx->current_meta_string);
+         }
+      }
+   }
+
+   pCtx->current_entry_val = NULL;
+
+   return JSON_Parser_Continue;
+}
+
+static JSON_Parser_HandlerResult JSONObjectMemberHandler(JSON_Parser parser, char *pValue, size_t length, JSON_StringAttributes attributes)
+{
+   JSONContext *pCtx = (JSONContext*)JSON_Parser_GetUserData(parser);
+   (void)attributes; /* unused */
+
+   if (pCtx->in_items && pCtx->object_depth == 2)
+   {
+      if (pCtx->array_depth == 1)
+      {
+         if (pCtx->current_entry_val)
+         {
+            /* something went wrong */
+            RARCH_WARN("JSON parsing failed at line %d.\n", __LINE__);
+            return JSON_Parser_Abort;
+         }
+
+         if (length)
+         {
+            if (string_is_equal(pValue, "path"))
+               pCtx->current_entry_val = &pCtx->current_entry->path;
+            else if (string_is_equal(pValue, "label"))
+               pCtx->current_entry_val = &pCtx->current_entry->label;
+            else if (string_is_equal(pValue, "core_path"))
+               pCtx->current_entry_val = &pCtx->current_entry->core_path;
+            else if (string_is_equal(pValue, "core_name"))
+               pCtx->current_entry_val = &pCtx->current_entry->core_name;
+            else if (string_is_equal(pValue, "crc32"))
+               pCtx->current_entry_val = &pCtx->current_entry->crc32;
+            else if (string_is_equal(pValue, "db_name"))
+               pCtx->current_entry_val = &pCtx->current_entry->db_name;
+            else
+            {
+               /* ignore unknown members */
+            }
+         }
+      }
+   }
+   else if (pCtx->object_depth == 1)
+   {
+      if (pCtx->array_depth == 0)
+      {
+         if (length)
+         {
+            pCtx->current_meta_string = strdup(pValue);
+         }
+      }
+   }
+
+   return JSON_Parser_Continue;
+}
+
 static bool playlist_read_file(
       playlist_t *playlist, const char *path)
 {
    unsigned i;
-   char buf[PLAYLIST_ENTRIES][1024];
-   intfstream_t *file = intfstream_open_file(
-         path, RETRO_VFS_FILE_ACCESS_READ,
-         RETRO_VFS_FILE_ACCESS_HINT_NONE);
+   bool new_format = true;
+   RFILE *file = filestream_open(path,
+         RETRO_VFS_FILE_ACCESS_READ, RETRO_VFS_FILE_ACCESS_HINT_NONE);
 
    /* If playlist file does not exist,
     * create an empty playlist instead.
     */
    if (!file)
       return true;
 
-   for (i = 0; i < PLAYLIST_ENTRIES; i++)
-      buf[i][0] = '\0';
+   /* Detect format of playlist */
+   {
+      char buf[16] = {0};
+      int64_t bytes_read = filestream_read(file, buf, 15);
+
+      filestream_seek(file, 0, SEEK_SET);
 
-   for (playlist->size = 0; playlist->size < playlist->cap; )
+      if (bytes_read == 15)
+      {
+         if (string_is_equal(buf, "{\n  \"version\": "))
```

Comment:
The detection between the differents formats while using the same file extension is the only issue I have with this. Shouldn't be that much of a problem though, as long as we make sure `{\n  "version": ` is always first in the new playlists' contents.

Leaving it to @twinaphex to do the final review! Nice work, bparker06. Excited to see how we can expand the playlist functionality through this.